lib/vrl/compiler/src/expression/query.rs

```diff
@@ -120,19 +120,11 @@ impl Expression for Query {
         use Target::*;
 
         match &self.target {
-            External => {
-                // `.` path must be an object
-                //
-                // TODO: make sure to enforce this
-                if self.path.is_root() {
-                    return TypeDef::object(Collection::any()).infallible();
-                }
-
-                match state.1.target() {
-                    None => TypeDef::any().infallible(),
-                    Some(details) => details.clone().type_def.at_path(&self.path.to_lookup()),
-                }
-            }
+            External => match state.1.target() {
+                None if self.path.is_root() => TypeDef::object(Collection::any()).infallible(),
+                None => TypeDef::any().infallible(),
```

Comment:
I left this in because I didn't think @fuchsnj's PR of allowing the root to be any type hadn't landed yet, but it seems that it did.

Did you intentionally not touch this code, or missed it? I guess we'll want to remove the `is_root` variant in this list now, correct?

Reply:
I missed it. I assumed a few things like this would get missed until I finish more of the work and do full testing of it.

Reply:
Gotcha. I'll remove the conditional above, and type-def it to `any` if no information is known about the root of the target.

Reply:
I backed out of this, as it would be a breaking change that exposed itself through a bunch of VRL integration tests that started failing.

We can still move forward with this, but we'd have to use schema support to inject the proper root type into the VRL compiler. Nothing we can't do, but also nothing I want to do in this fix PR.